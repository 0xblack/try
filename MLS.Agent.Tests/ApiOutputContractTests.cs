using System.Threading.Tasks;
using Assent;
using Recipes;
using WorkspaceServer.Models;
using WorkspaceServer.Models.Execution;
using WorkspaceServer.Tests;
using Xunit;
using Xunit.Abstractions;

namespace MLS.Agent.Tests
{
    public class ApiOutputContractTests : ApiViaHttpTestsBase
    {
        private readonly Configuration configuration;

        public ApiOutputContractTests(ITestOutputHelper output) : base(output)
        {
            configuration = new Configuration()
                .UsingExtension("json");

#if !DEBUG
            configuration = configuration.SetInteractive(false);
#endif
        }

        [Fact]
        public async Task The_Run_contract_for_compiling_code_has_not_been_broken()
        {
            var viewport = ViewportCode();

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id);

            var response = await CallRun(requestJson.ToJson());

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_Run_contract_for_noncompiling_code_has_not_been_broken()
        {
            var viewport = ViewportCode("doesn't compile");

            var request = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id);

            var requestBody = request.ToJson();

            var response = await CallRun(requestBody);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_Completions_contract_has_not_been_broken()
        {
            var viewport = ViewportCode("Console.Ou$$");

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id).ToJson();

            var response = await CallCompletion(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_signature_help_contract_has_not_been_broken()
        {
            var viewport = ViewportCode("Console.Write($$");

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode(),
                        viewport
                    }),
                activeBufferId: viewport.Id).ToJson();

            var response = await CallSignatureHelp(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_instrumentation_contract_has_not_been_broken()
        {

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode("int a = 1; int b = 2; a = 3; b = a;")
                    },
                    includeInstrumentation: true)
            ).ToJson();

            var response = await CallRun(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        [Fact]
        public async Task The_run_contract_with_no_instrumentation_has_not_been_broken()
        {

            var requestJson = new WorkspaceRequest(
                new Workspace(
                    workspaceType: "console",
                    buffers: new[]
                    {
                        EntrypointCode("int a = 1; int b = 2; a = 3; b = a;")
                    },
                    includeInstrumentation: false)
            ).ToJson();

            var response = await CallRun(requestJson);

            var result = await response.Content.ReadAsStringAsync();

            this.Assent(result.FormatJson(), configuration);
        }

        private static Workspace.Buffer EntrypointCode(string mainContent = @"Console.WriteLine(Sample.Method());$$")
        {
            var input = $@"
using System;
using System.Linq;

namespace Example
{{
    public class Program
    {{
        public static void Main()
        {{
            {mainContent}
        }}       
    }}
}}".EnforceLF();
                
            MarkupTestFile.GetPosition(input, out string output, out var position);

            return new Workspace.Buffer(
                "EntrypointCode.cs",
                output,
                position ?? 0);
        }

        private static Workspace.Buffer ViewportCode(string methodContent = @"return ""Hello world!"";$$ ")
        {
            var input = $@"
using System.Collections.Generic;
using System;

namespace Example
{{
    public static class Sample
    {{
        public static object Method()
        {{
#region viewport
            {methodContent}
#endregion
        }}
    }}
}}".EnforceLF();

            MarkupTestFile.GetPosition(input, out string output, out var position);

            return new Workspace.Buffer(
                "ViewportCode.cs",
                output,
                position ?? 0);
        }
    }
}