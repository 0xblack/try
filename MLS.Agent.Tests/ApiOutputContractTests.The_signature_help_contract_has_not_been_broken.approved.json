{
  "signatures": [
    {
      "name": "Write",
      "label": "void Console.Write(bool value)",
      "documentation": "\n      <summary>Writes the text representation of the specified Boolean value to the standard output stream.</summary>\n      <param name=\"value\">The value to write.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "value",
          "label": "bool value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(char value)",
      "documentation": "\n      <summary>Writes the specified Unicode character value to the standard output stream.</summary>\n      <param name=\"value\">The value to write.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "value",
          "label": "char value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(char[] buffer)",
      "documentation": "\n      <summary>Writes the specified array of Unicode characters to the standard output stream.</summary>\n      <param name=\"buffer\">A Unicode character array.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "buffer",
          "label": "char[] buffer",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(char[] buffer, int index, int count)",
      "documentation": "\n      <summary>Writes the specified subarray of Unicode characters to the standard output stream.</summary>\n      <param name=\"buffer\">An array of Unicode characters.</param>\n      <param name=\"index\">The starting position in buffer.</param>\n      <param name=\"count\">The number of characters to write.</param>\n      <exception cref=\"T:System.ArgumentNullException\"><paramref name=\"buffer\">buffer</paramref> is null.</exception>\n      <exception cref=\"T:System.ArgumentOutOfRangeException\"><paramref name=\"index\">index</paramref> or <paramref name=\"count\">count</paramref> is less than zero.</exception>\n      <exception cref=\"T:System.ArgumentException\"><paramref name=\"index\">index</paramref> plus <paramref name=\"count\">count</paramref> specify a position that is not within <paramref name=\"buffer\">buffer</paramref>.</exception>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "buffer",
          "label": "char[] buffer",
          "documentation": ""
        },
        {
          "name": "index",
          "label": "int index",
          "documentation": ""
        },
        {
          "name": "count",
          "label": "int count",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(decimal value)",
      "documentation": "\n      <summary>Writes the text representation of the specified <see cref=\"T:System.Decimal\"></see> value to the standard output stream.</summary>\n      <param name=\"value\">The value to write.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "value",
          "label": "decimal value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(double value)",
      "documentation": "\n      <summary>Writes the text representation of the specified double-precision floating-point value to the standard output stream.</summary>\n      <param name=\"value\">The value to write.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "value",
          "label": "double value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(int value)",
      "documentation": "\n      <summary>Writes the text representation of the specified 32-bit signed integer value to the standard output stream.</summary>\n      <param name=\"value\">The value to write.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "value",
          "label": "int value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(long value)",
      "documentation": "\n      <summary>Writes the text representation of the specified 64-bit signed integer value to the standard output stream.</summary>\n      <param name=\"value\">The value to write.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "value",
          "label": "long value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(object value)",
      "documentation": "\n      <summary>Writes the text representation of the specified object to the standard output stream.</summary>\n      <param name=\"value\">The value to write, or null.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "value",
          "label": "object value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(float value)",
      "documentation": "\n      <summary>Writes the text representation of the specified single-precision floating-point value to the standard output stream.</summary>\n      <param name=\"value\">The value to write.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "value",
          "label": "float value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(string value)",
      "documentation": "\n      <summary>Writes the specified string value to the standard output stream.</summary>\n      <param name=\"value\">The value to write.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "value",
          "label": "string value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(string format, object arg0)",
      "documentation": "\n      <summary>Writes the text representation of the specified object to the standard output stream using the specified format information.</summary>\n      <param name=\"format\">A composite format string.</param>\n      <param name=\"arg0\">An object to write using format.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n      <exception cref=\"T:System.ArgumentNullException\"><paramref name=\"format\">format</paramref> is null.</exception>\n      <exception cref=\"T:System.FormatException\">The format specification in <paramref name=\"format\">format</paramref> is invalid.</exception>\n    ",
      "parameters": [
        {
          "name": "format",
          "label": "string format",
          "documentation": ""
        },
        {
          "name": "arg0",
          "label": "object arg0",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(string format, object arg0, object arg1)",
      "documentation": "\n      <summary>Writes the text representation of the specified objects to the standard output stream using the specified format information.</summary>\n      <param name=\"format\">A composite format string.</param>\n      <param name=\"arg0\">The first object to write using format.</param>\n      <param name=\"arg1\">The second object to write using format.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n      <exception cref=\"T:System.ArgumentNullException\"><paramref name=\"format\">format</paramref> is null.</exception>\n      <exception cref=\"T:System.FormatException\">The format specification in <paramref name=\"format\">format</paramref> is invalid.</exception>\n    ",
      "parameters": [
        {
          "name": "format",
          "label": "string format",
          "documentation": ""
        },
        {
          "name": "arg0",
          "label": "object arg0",
          "documentation": ""
        },
        {
          "name": "arg1",
          "label": "object arg1",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(string format, object arg0, object arg1, object arg2)",
      "documentation": "\n      <summary>Writes the text representation of the specified objects to the standard output stream using the specified format information.</summary>\n      <param name=\"format\">A composite format string.</param>\n      <param name=\"arg0\">The first object to write using format.</param>\n      <param name=\"arg1\">The second object to write using format.</param>\n      <param name=\"arg2\">The third object to write using format.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n      <exception cref=\"T:System.ArgumentNullException\"><paramref name=\"format\">format</paramref> is null.</exception>\n      <exception cref=\"T:System.FormatException\">The format specification in <paramref name=\"format\">format</paramref> is invalid.</exception>\n    ",
      "parameters": [
        {
          "name": "format",
          "label": "string format",
          "documentation": ""
        },
        {
          "name": "arg0",
          "label": "object arg0",
          "documentation": ""
        },
        {
          "name": "arg1",
          "label": "object arg1",
          "documentation": ""
        },
        {
          "name": "arg2",
          "label": "object arg2",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(string format, params object[] arg)",
      "documentation": "\n      <summary>Writes the text representation of the specified array of objects to the standard output stream using the specified format information.</summary>\n      <param name=\"format\">A composite format string.</param>\n      <param name=\"arg\">An array of objects to write using format.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n      <exception cref=\"T:System.ArgumentNullException\"><paramref name=\"format\">format</paramref> or <paramref name=\"arg\">arg</paramref> is null.</exception>\n      <exception cref=\"T:System.FormatException\">The format specification in <paramref name=\"format\">format</paramref> is invalid.</exception>\n    ",
      "parameters": [
        {
          "name": "format",
          "label": "string format",
          "documentation": ""
        },
        {
          "name": "arg",
          "label": "params object[] arg",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(uint value)",
      "documentation": "\n      <summary>Writes the text representation of the specified 32-bit unsigned integer value to the standard output stream.</summary>\n      <param name=\"value\">The value to write.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "value",
          "label": "uint value",
          "documentation": ""
        }
      ]
    },
    {
      "name": "Write",
      "label": "void Console.Write(ulong value)",
      "documentation": "\n      <summary>Writes the text representation of the specified 64-bit unsigned integer value to the standard output stream.</summary>\n      <param name=\"value\">The value to write.</param>\n      <exception cref=\"T:System.IO.IOException\">An I/O error occurred.</exception>\n    ",
      "parameters": [
        {
          "name": "value",
          "label": "ulong value",
          "documentation": ""
        }
      ]
    }
  ],
  "activeSignature": 0,
  "activeParameter": 0
}